torch/distributed/tensor/parallel/style.py

```diff
@@ -195,32 +193,51 @@ def _prepare_input_fn(input_layouts, desired_input_layouts, inputs, device_mesh)
             input_tensor = input_tensor.redistribute(placements=desired_input_layouts)
         return input_tensor
 
-    def _partition_fn(self, name, module, device_mesh):
-        if isinstance(module, nn.Linear):
-            # Rowwise shard weight to Shard(1), bias to Replicate(), weight be Shard(1)
-            # means Rowwise as Linear is input * weight^T + bias, where
-            # weight would become Shard(0)
-            module.register_parameter("weight", nn.Parameter(
-                distribute_tensor(module.weight, device_mesh, [Shard(1)])
+    def _partition_linear_fn(self, name, module, device_mesh):
+        # Rowwise shard weight to Shard(1), bias to Replicate(), weight be Shard(1)
+        # means Rowwise as nn.Linear is input * weight^T + bias, where
+        # weight would become Shard(0)
+        module.register_parameter("weight", nn.Parameter(
+            distribute_tensor(module.weight, device_mesh, [Shard(1)])
+        ))
+        if module.bias is not None:
+            module.register_parameter("bias", nn.Parameter(
+                distribute_tensor(module.bias, device_mesh, [Replicate()])
             ))
-            if module.bias is not None:
-                module.register_parameter("bias", nn.Parameter(
-                    distribute_tensor(module.bias, device_mesh, [Replicate()])
-                ))
-        else:
-            raise NotImplementedError("RowwiseParallel currently only support nn.Linear!")
+
+    def _partition_embedding_fn(self, name, module, device_mesh):
+        # rowwise shard embedding.weight is Shard(0)
+        for name, param in module.named_parameters():
+            dist_param = nn.Parameter(
+                distribute_tensor(param, device_mesh, [Shard(0)])
+            )
+            module.register_parameter(name, dist_param)
 
     @staticmethod
     def _prepare_output_fn(output_layouts, use_local_output, outputs, device_mesh):
+        # Rowwise sharding produces partial output, depending on output layouts:
+        # 1. to replicate -> allreduce
+        # 2. to shard -> reduce_scatter
         outputs = outputs.redistribute(placements=output_layouts)
         # back to local tensor if use_local_output is True
         return outputs.to_local() if use_local_output else outputs
 
     def _apply(self, module: nn.Module, device_mesh: DeviceMesh) -> nn.Module:
+        if isinstance(module, nn.Linear):
+            partition_fn = self._partition_linear_fn
+            # rowwise linear runtime sharding requires input tensor shard on last dim
+            self.desired_input_layouts: Tuple[Placement, ...] = (Shard(-1), )
+        elif isinstance(module, nn.Embedding):
+            partition_fn = self._partition_embedding_fn
+            # rowwise embedding runtime sharding requires input tensor replicated
+            self.desired_input_layouts = (Replicate(), )
```

Comment:
up to you whether or not we want L229 and L233 more symmetric

Reply:
If you are referring to type annotation then mypy would complain if I declare the type twice.